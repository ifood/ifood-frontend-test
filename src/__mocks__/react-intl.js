import React from 'react';

function useIntl() {
  return {
    formatMessage: jest.fn().mockImplementation(({ id }) => `intl::${id}`),
  };
}

const FormattedMessage = () => null;

/* eslint-disable-next-line */
const IntlProvider = ({ children }) => <div>{children}</div>;

export { useIntl, FormattedMessage, IntlProvider };
