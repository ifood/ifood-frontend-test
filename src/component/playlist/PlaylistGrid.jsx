/* eslint-disable react-hooks/exhaustive-deps, camelcase */
import React, { useCallback, useEffect, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
  Accordion,
  IconButton,
  TextField,
  Typography,
} from '@material-ui/core';
import { FilterList } from '@material-ui/icons';
import GridList from '../gridlist/GridList';
import PlaylistCard from './PlaylistCard';
import { CenteredFlex, LayoutContent } from '../containers/Containers';
import FilterForm from '../filter/FilterForm';
import PlaylistError from './PlaylistError';

const Content = styled(LayoutContent)`
  margin-top: 150px;
`;
const FilterContainer = styled.div`
  display: flex;
  padding-bottom: 16px;
`;
const AccordionContainer = styled.div`
  padding-bottom: 16px;
`;

function itemRenderer(element) {
  return <PlaylistCard playlist={element} />;
}

function PlaylistGrid(props) {
  const {
    data,
    errorMessage,
    filterData,
    filters,
    onChangeFilter,
    title,
  } = props;
  const [filterText, setFilterText] = useState('');
  const [gridData, setGridData] = useState([]);
  const [expandedFilter, setExpandedFilter] = useState(false);

  useEffect(() => {
    setGridData([...data]);
  }, [data]);

  const handleChangeTextFilter = useCallback((element) => {
    setFilterText(element?.target?.value);
  }, []);

  const handleExpandFilter = useCallback(() => {
    setExpandedFilter(!expandedFilter);
  }, [expandedFilter]);

  useEffect(() => {
    const filterValues = data.filter(
      (element) =>
        element.name.toUpperCase().includes(filterText.toUpperCase()) ||
        element.description.toUpperCase().includes(filterText.toUpperCase()),
    );

    setGridData(filterValues);
  }, [filterText]);

  const renderGrid = useMemo(() => {
    if (errorMessage != null) {
      return <PlaylistError error={errorMessage} />;
    }

    if (gridData.length > 0) {
      return (
        <>
          <Typography color="textPrimary" variant="h5" gutterBottom>
            {title}
          </Typography>
          <GridList
            data={gridData}
            itemRenderer={itemRenderer}
            filters={filters}
          />
        </>
      );
    }

    return (
      <CenteredFlex>
        <Typography color="textPrimary" variant="h5" gutterBottom>
          Your search result did not return any playlist
        </Typography>
      </CenteredFlex>
    );
  }, [title, gridData, itemRenderer, filters, errorMessage]);

  return (
    <Content>
      <AccordionContainer>
        <Accordion expanded={expandedFilter}>
          <FilterContainer>
            <TextField
              id="local-filter"
              label="Search playlist"
              variant="filled"
              onChange={handleChangeTextFilter}
              fullWidth
            />
            <IconButton color="primary" onClick={handleExpandFilter}>
              <FilterList />
            </IconButton>
          </FilterContainer>
          <FilterForm
            filters={filters}
            onChange={onChangeFilter}
            filterData={filterData}
          />
        </Accordion>
      </AccordionContainer>
      {renderGrid}
    </Content>
  );
}

PlaylistGrid.defaultProps = {
  data: [],
};

PlaylistGrid.propTypes = {
  data: PropTypes.array,
  title: PropTypes.string,
  errorMessage: PropTypes.string,
  filters: PropTypes.array,
  filterData: PropTypes.object,
  onChangeFilter: PropTypes.func,
};

export default PlaylistGrid;
