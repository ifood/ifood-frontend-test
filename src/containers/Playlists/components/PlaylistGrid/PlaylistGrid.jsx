
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import GridList from '@material-ui/core/GridList';
import GridListTile from '@material-ui/core/GridListTile';
import GridListTileBar from '@material-ui/core/GridListTileBar';

const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  margin-top: 25px;
`;

const Link = styled.a`
  text-decoration: none;
  display: block;
`;

const StyledGridListTileBar = styled(GridListTileBar)`
  max-width: 300px;
`;

class PlaylistGrid extends Component {
  constructor(props) {
    super(props);
    this.state = {
      columns: 0,
    };
  }

  componentDidMount() {
    this.evaluateColumnsNumber();
    window.addEventListener('resize', this.evaluateColumnsNumber);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.evaluateColumnsNumber);
  }

  evaluateColumnsNumber = () => {
    this.setState({
      columns: Math.floor(window.innerWidth / 300) || 1,
    });
  }

  render() {
    const { playlists } = this.props;
    const { columns } = this.state;
    const cols = columns <= playlists.length ? columns : playlists.length;

    return (
      <Container>
        <GridList cellHeight={300} cols={cols}>
          {playlists.map(playlist => (
            <GridListTile key={playlist.images[0].url} cols={1}>
              <Link href={playlist.external_urls.spotify} target="_blank">
                <img src={playlist.images[0].url} alt="playlist" />
                <StyledGridListTileBar
                  title={playlist.name}
                  subtitle={playlist.owner.display_name}
                />
              </Link>
            </GridListTile>
          ))}
        </GridList>
      </Container>
    );
  }
}

PlaylistGrid.propTypes = {
  playlists: PropTypes.array,
};

PlaylistGrid.defaultProps = {
  playlists: [],
};

export default PlaylistGrid;
