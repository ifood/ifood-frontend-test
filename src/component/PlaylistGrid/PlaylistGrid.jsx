import React from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import Playlist from '../Playlist';
import Info from '../Info';
import Grid from '../Grid';

function itemRenderer(playlist) {
  return <Playlist key={playlist?.id} playlist={playlist} />;
}

function PlaylistGrid({ title, items }) {
  const intl = useIntl();

  const gridTitle =
    title != null ? title : intl.formatMessage({ id: 'featured.playlists' });

  if (items == null || items.length === 0) {
    return (
      <Info
        icon="search"
        title={intl.formatMessage({ id: 'no.playlist' })}
        info={intl.formatMessage({ id: 'choose.filters.message' })}
      />
    );
  }

  return <Grid title={gridTitle} items={items} itemRenderer={itemRenderer} />;
}

PlaylistGrid.propTypes = {
  title: PropTypes.string,
  items: PropTypes.array,
};

export default React.memo(PlaylistGrid);
