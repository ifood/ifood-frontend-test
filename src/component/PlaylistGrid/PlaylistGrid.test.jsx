import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { useIntl } from 'react-intl';
import PlaylistGrid from './PlaylistGrid';

const intl = useIntl();
const items = [
  {
    id: 'id',
  },
];

describe('<PlaylistGrid />', () => {
  it('snapshot with items', () => {
    const playlistGrid = mount(<PlaylistGrid title="Grid" items={items} />);
    jestExpect(toJson(playlistGrid)).toMatchSnapshot();
  });

  it('snapshot without items', () => {
    const playlistGrid = mount(<PlaylistGrid />);
    jestExpect(toJson(playlistGrid)).toMatchSnapshot();
  });

  it('should propagate title to Grid if provided', () => {
    const title = 'Grid';
    const playlistGrid = shallow(<PlaylistGrid title={title} items={items} />);
    const gridTitle = playlistGrid.get(0).props.title;
    expect(gridTitle).to.be.equal(title);
  });

  it('should use a standard title if the title is not provided', () => {
    const title = intl.formatMessage({ id: 'featured.playlists' });
    const playlistGrid = shallow(<PlaylistGrid items={items} />);
    const gridTitle = playlistGrid.get(0).props.title;
    expect(gridTitle).to.be.equal(title);
  });
});
