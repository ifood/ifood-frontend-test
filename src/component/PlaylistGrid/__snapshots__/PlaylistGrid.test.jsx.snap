// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaylistGrid /> snapshot with items 1`] = `
<Memo(PlaylistGrid)
  items={
    Array [
      Object {
        "id": "id",
      },
    ]
  }
  title="Grid"
>
  <Memo(Grid)
    itemRenderer={[Function]}
    items={
      Array [
        Object {
          "id": "id",
        },
      ]
    }
    title="Grid"
  >
    <h1>
      Grid
    </h1>
    <styled.div
      columnMinWidth="200px"
    >
      <div
        className="sc-AxiKw ixJNhB"
      >
        <Playlist
          key="id"
          playlist={
            Object {
              "id": "id",
            }
          }
        >
          <Styled(Card)
            fluid={true}
            onClick={[Function]}
          >
            <Card
              className="sc-AxjAm ihGEoy"
              fluid={true}
              onClick={[Function]}
            >
              <a
                className="ui fluid card sc-AxjAm ihGEoy"
                onClick={[Function]}
              >
                <Styled(Image)
                  src="DefaultPlaylistImage.jpg"
                  ui={false}
                  wrapped={true}
                >
                  <Image
                    as="img"
                    className="sc-AxirZ fGQfeF"
                    src="DefaultPlaylistImage.jpg"
                    ui={false}
                    wrapped={true}
                  >
                    <div
                      className="image sc-AxirZ fGQfeF"
                    >
                      <img
                        src="DefaultPlaylistImage.jpg"
                      />
                    </div>
                  </Image>
                </Styled(Image)>
                <CardContent>
                  <div
                    className="content"
                  >
                    <CardHeader>
                      <div
                        className="header"
                      />
                    </CardHeader>
                    <CardDescription>
                      <div
                        className="description"
                      >
                        <span
                          dangerouslySetInnerHTML={
                            Object {
                              "__html": undefined,
                            }
                          }
                        />
                      </div>
                    </CardDescription>
                  </div>
                </CardContent>
              </a>
            </Card>
          </Styled(Card)>
        </Playlist>
      </div>
    </styled.div>
  </Memo(Grid)>
</Memo(PlaylistGrid)>
`;

exports[`<PlaylistGrid /> snapshot without items 1`] = `
<Memo(PlaylistGrid)>
  <Info
    icon="search"
    info="intl::choose.filters.message"
    title="intl::no.playlist"
  >
    <Segment
      placeholder={true}
    >
      <div
        className="ui placeholder segment"
      >
        <Header
          icon={true}
        >
          <div
            className="ui icon header"
          >
            <Icon
              as="i"
              name="search"
            >
              <i
                aria-hidden="true"
                className="search icon"
                onClick={[Function]}
              />
            </Icon>
            intl::no.playlist
          </div>
        </Header>
        <SegmentInline>
          <div
            className="inline"
          >
            intl::choose.filters.message
          </div>
        </SegmentInline>
      </div>
    </Segment>
  </Info>
</Memo(PlaylistGrid)>
`;
